src/request/providers/xhr.ts

```diff
@@ -0,0 +1,310 @@
+	}
+
+	blob(): Task<Blob> {
+		return Task.reject<Blob>(new Error('Blob not supported'));
+	}
+
+	formData(): Task<FormData> {
+		return Task.reject<FormData>(new Error('FormData not supported'));
+	}
+
+	text(): Task<string> {
+		return <any> getDataTask(this).then((request: XMLHttpRequest) => {
+			return String(request.responseText);
+		});
+	}
+
+	xml(): Task<Document> {
+		return <any> this.text().then((text: string) => {
+			const parser = new DOMParser();
+			return parser.parseFromString(text, this.headers.get('content-type') || 'text/html');
+		});
+	}
+}
+
+if (has('blob')) {
+	XhrResponse.prototype.blob = function (this: XhrResponse): Task<Blob> {
+		return <any> getDataTask(this).then((request: XMLHttpRequest) => request.response);
+	};
+
+	XhrResponse.prototype.text = function (this: XhrResponse): Task<string> {
+		return <any> this.blob().then(getTextFromBlob);
+	};
+
+	if (has('arraybuffer')) {
+		XhrResponse.prototype.arrayBuffer = function (this: XhrResponse): Task<ArrayBuffer> {
+			return <any> this.blob().then(getArrayBufferFromBlob);
+		};
+	}
+}
+
+if (has('formdata')) {
+	XhrResponse.prototype.formData = function (this: XhrResponse): Task<FormData> {
+		return <any> this.text().then((text: string) => {
+			const data = new FormData();
+
+			text.trim().split('&').forEach(keyValues => {
+				if (keyValues) {
+					const pairs = keyValues.split('=');
+					const name = (pairs.shift() || '').replace(/\+/, ' ');
+					const value = pairs.join('=').replace(/\+/, ' ');
+
+					data.append(decodeURIComponent(name), decodeURIComponent(value));
+				}
+			});
+
+			return data;
+		});
+	};
+}
+
+function noop () {}
+
+function setOnError(request: XMLHttpRequest, reject: Function) {
+	request.onerror = function (event) {
+		reject(new TypeError(event.error || 'Network request failed'));
+	};
+}
+
+export default function xhr(url: string, options: XhrRequestOptions = {}): Task<Response> {
+	const request = new XMLHttpRequest();
+	const requestUrl = generateRequestUrl(url, options);
+
+	options = Object.create(options);
+
+	if (!options.method) {
+		options.method = 'GET';
+	}
+
+	if ((!options.user || !options.password) && options.auth) {
+		const auth = options.auth.split(':');
+		options.user = decodeURIComponent(auth[ 0 ]);
+		options.password = decodeURIComponent(auth[ 1 ]);
+	}
+
+	let isAborted = false;
+
+	function abort() {
+		isAborted = true;
+		if (request) {
+			request.abort();
+			request.onreadystatechange = noop;
+		}
+	}
+
+	let timeoutHandle: Handle;
+	let timeoutReject: Function;
+
+	const task = new Task<Response>((resolve, reject) => {
+		timeoutReject = reject;
+
+		request.onreadystatechange = function () {
+			if (isAborted) {
+				return;
+			}
+
+			if (request.readyState === 2) {
+				const response = new XhrResponse(request);
+
+				const task = new Task<XMLHttpRequest>((resolve, reject) => {
+					timeoutReject = reject;
+
+					request.onprogress = function (event: any) {
```

Comment:
This only handles download progress. It would be nice (perhaps in another PR) to handle upload progress as well.